src/s3_fetchers.rs

```diff
@@ -1,84 +1,105 @@
-    let folder_names = match response.common_prefixes() {
+    Ok(match response.common_prefixes {
         None => vec![],
         Some(common_prefixes) => common_prefixes
             .into_iter()
-            .filter_map(|common_prefix| common_prefix.prefix.clone())
+            .filter_map(|common_prefix| common_prefix.prefix)
             .collect(),
-    };
-
-    Ok(crate::types::ListObjectResponse {
-        continuation_token,
-        folder_names,
     })
 }
 
-/// By the given block height (`key`) gets the objects:
+/// By the given block height (`block_height_prefix`) gets the objects:
 /// - block.json
 /// - shard_N.json
-/// Reads the content of the objects and parses it to JSON.
-/// Returns the result in a single JSON
-pub(crate) async fn get_object(
-    client: &Client,
-    bucket: &str,
-    key: &str,
-    tracked_shards: &Vec<u8>,
-) -> Result<serde_json::Value, Error> {
-    let mut main_json = serde_json::json!({});
-    let block_json = {
-        let response = client
-            .get_object()
-            .bucket(bucket)
-            .key(format!("{}block.json", key))
-            .send()
-            .await?;
+/// Reads the content of the objects and parses as a JSON.
+/// Returns the result in `near_indexer_primitives::StreamerMessage`
+pub(crate) async fn fetch_streamer_message(
+    s3_client: &Client,
+    s3_bucket_name: &str,
+    block_height_prefix: &str,
+) -> anyhow::Result<near_indexer_primitives::StreamerMessage> {
+    let block_view = {
+        let response = loop {
+            if let Ok(response) = s3_client
+                .get_object()
+                .bucket(s3_bucket_name)
+                .key(format!("{}block.json", block_height_prefix))
+                .send()
+                .await
+            {
+                break response;
+            }
+        };
 
         let body_bytes = response.body.collect().await.unwrap().into_bytes();
 
-        serde_json::from_slice(body_bytes.as_ref()).unwrap()
+        serde_json::from_slice::<near_indexer_primitives::views::BlockView>(body_bytes.as_ref())
+            .unwrap()
     };
 
-    main_json["block"] = block_json;
+    let shards_num: u64 = block_view.header.chunks_included;
 
-    let mut shards: Vec<serde_json::Value> = vec![];
-    // TODO: undefined bahaviour in case if we track N > 1 shards
-    // but there is only N-1 shards, we need to handle it. Probably,
-    // by checking the `block.chunks` amount
-    for shard_id in tracked_shards {
-        let response = client
-            .get_object()
-            .bucket(bucket)
-            .key(format!("{}shard_{}.json", key, shard_id))
-            .send()
-            .await?;
+    let mut shards: Vec<near_indexer_primitives::IndexerShard> = vec![];
 
-        let body_bytes = response.body.collect().await.unwrap().into_bytes();
+    let mut shards_futures: futures::stream::FuturesOrdered<_> = (0..shards_num)
```

Comment:
It seems this can be simplified using [`StreamExt::collect`](https://docs.rs/futures/0.2.1/futures/trait.StreamExt.html#method.collect) to:

```rust
    let shards: Vec<near_indexer_primitives::IndexerShard> = (0..block_view.header.chunks_included)
        .collect::<Vec<u64>>()
        .into_iter()
        .map(|shard_id| {
            fetch_shard_or_retry(s3_client, s3_bucket_name, block_height_prefix, shard_id)
        })
        .collect::<futures::stream::FuturesOrdered<_>>()
        .collect()
        .await;
```